x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java

```diff
@@ -108,144 +133,214 @@ protected void masterOperation(
-                                        if (resizeResponse.isAcknowledged()) {
-                                            // 6.
-                                            publishMetadata(originalIndexName, tmpIndexName, rollupIndexName, listener);
+                                    // 5. Refresh rollup index
+                                    refreshIndex(rollupIndexName, parentTask, ActionListener.wrap(refreshIndexResponse -> {
+                                        if (refreshIndexResponse.getFailedShards() == 0) {
+                                            // 6. Add rollup index to data stream and publish rollup metadata
+                                            updateRollupMetadata(sourceIndexName, rollupIndexName, request, ActionListener.wrap(resp -> {
+                                                if (resp.isAcknowledged()) {
+                                                    // 9. Force-merge the rollup index to a single segment
+                                                    forceMergeIndex(
+                                                        rollupIndexName,
+                                                        parentTask,
+                                                        ActionListener.wrap(
+                                                            mergeIndexResp -> listener.onResponse(AcknowledgedResponse.TRUE),
+                                                            e -> listener.onFailure(
+                                                                new ElasticsearchException(
+                                                                    "Failed to force-merge index [" + rollupIndexName + "]",
+                                                                    e
+                                                                )
+                                                            )
+                                                        )
+                                                    );
+                                                } else {
+                                                    deleteRollupIndex(
+                                                        sourceIndexName,
+                                                        rollupIndexName,
+                                                        parentTask,
+                                                        listener,
+                                                        new ElasticsearchException(
+                                                            "Failed to publish new cluster state with rollup metadata"
+                                                        )
+                                                    );
+                                                }
+                                            },
+                                                e -> deleteRollupIndex(
```

Comment:
I wonder if there is a good way to reduce the duplicate calls here. Oh well, in a follow up.